apps/gateway/test/champions_test.exs

```diff
@@ -0,0 +1,251 @@
+        response_type: {:unit_and_currencies, %UnitAndCurrencies{unit: unit, user_currency: [user_currency]}}
+      } = get_last_message()
+
+      assert unit.level == level + 1
+      assert user_currency.currency.name == "Gold"
+      assert user_currency.amount == gold - level_up_cost
+
+      # Cannot level up because unit is level 20 with tier 1
+      assert unit.level == 20
+      assert unit.tier == 1
+      :ok = SocketTester.level_up_unit(socket_tester, user.id, unit.id)
+      fetch_last_message(socket_tester)
+      assert_receive %WebSocketResponse{response_type: {:error, %Error{reason: "cant_level_up"}}}
+
+      #### Tier up
+
+      user_gold = Currencies.get_amount_of_currency_by_name(user.id, "Gold")
+      user_gems = Currencies.get_amount_of_currency_by_name(user.id, "Gems")
+
+      [%CurrencyCost{currency_id: _gold_id, amount: gold_cost}, %CurrencyCost{currency_id: _gems_id, amount: gems_cost}] =
+        Units.calculate_tier_up_cost(unit)
+
+      :ok = SocketTester.tier_up_unit(socket_tester, user.id, unit.id)
+      fetch_last_message(socket_tester)
+
+      %WebSocketResponse{
+        response_type: {:unit_and_currencies, %UnitAndCurrencies{unit: unit, user_currency: user_currencies}}
+      } = get_last_message()
+
+      user_currencies =
+        Enum.into(user_currencies, %{}, fn %UserCurrency{
+                                             currency: %Currency{name: name},
+                                             amount: amount
+                                           } ->
+          {name, amount}
+        end)
+
+      assert unit.level == 20
+      assert unit.tier == 2
+      assert user_currencies["Gold"] == user_gold - gold_cost
+      assert user_currencies["Gems"] == user_gems - gems_cost
+
+      # TODO: Check that we can now level up
+
+      #### Rank up (fuse)
+
+      {:ok, unit} =
+        GameBackend.Units.insert_unit(%{
+          user_id: user.id,
+          level: 220,
+          tier: 8,
+          rank: Units.get_rank(:star5),
+          selected: false,
+          character_id: muflus.id
+        })
+
+      unit = Repo.preload(unit, [:character])
+
+      rank = unit.rank
+      user_units_count = user.id |> GameBackend.Units.get_units() |> Enum.count()
+
+      # Add to-be-consumed units
+
+      {:ok, same_faction_character} =
+        GameBackend.Units.Characters.insert_character(%{
+          game_id: Utils.game_id(),
+          active: true,
+          name: "SameFactionUnit",
+          faction: muflus.faction,
+          quality: Champions.Units.get_quality(:rare)
+        })
+
+      # We will need three 5* Muflus and two i2 of the same faction
+      # For the same faction, we will do one of each unit.
+      units_to_consume = create_units_to_consume(user, muflus, same_faction_character)
+
+      # Characters need a certain quality to rank up
+      assert unit.character.quality == Units.get_quality(:epic)
+
+      # TODO: Check that we cant tier up again without ranking up
```

Comment:
I think we should add this on this pr since it's just a validation

Reply:
Sure

Reply:
Done!